File: Chapter4-RegressionofAContinuousVariable/cap4.r
#### CAPÍTULO 4 ####
# Algoritmos de ML relacionados a regressão:
# - Regressão simples
# - Regressão múltipla
# - Regressão polinomial
# - KNN

setwd("/Users/fernandovieira/OneDrive/1. Educacao/INFO E ESTAT/RAIZ_Livro_ML_Tinnian_Ganesh/PracticalMachineLearningWithRandPython/Chapter4-RegressionofAContinuousVariable")
dir()
source("RFunctions.R")

## Ler os dados
df <- read.csv("Boston.csv") # Considerando o pacote "RFunctions.R", o nome do data frame sempre deve ser 'df'
head(df)
tail(df)

names(df) # https://www.cs.toronto.edu/~delve/data/boston/bostonDetail.html
nrow(df)

### REGRESSÃO SIMPLES
## Dividir os dados entre dados de treino e de teste (75, 25)
treino_idx <- trainTestSplit(df, trainPercent=75, seed=5) # Função importada de "RFunctions.R"
treino_idx # não entendi de qual colunas se extraiu os dados
length(treino_idx)

treino <- df[treino_idx,]
teste <- df[-treino_idx,]

dim(treino)
dim(teste)

## Regressão linear 
# y: Median value of owner-occupied homes in $1000's
# x: lower status of the population
linreg <- lm(medv~lstat, data=df)
summary(linreg, digits = 2)

plot(df$lstat,df$medv, xlab="Lower status (%)",ylab="Median value of owned homes ($1000)", main="Median value of homes ($1000) vs Lowe status (%)")
abline(linreg,lwd=3,col="red")

rsquared <- Rsquared(linreg,teste,teste$medv)
rsquared

round(Rsquared(linreg,treino,treino$medv), 4)

### REGRESSÃO MÚLTIPLA
## Ler os dados
crimesDF <- read.csv("crimes.csv")
head(crimesDF)
tail(crimesDF)

names(crimesDF) # https://rpubs.com/ksakwa/470184

# Remover as 7 primeiras colunas
crimesDF1 <- crimesDF[,7:length(crimesDF)]

# Converter todos para numérico
crimesDF2 <- sapply(crimesDF1, as.numeric)
head(crimesDF2)
tail(crimesDF2)

# Procurar por NAs
a <- is.na(crimesDF2)
a

# Substituir NAs por 0
crimesDF2[a] <- 0

# Criar como dataframe
crimesDF2 <- as.data.frame(crimesDF2)

# Separar treino e teste
treino_idx <- trainTestSplit(crimesDF2, trainPercent=75, seed=5)
treino <- crimesDF2[treino_idx,]
teste <- crimesDF2[-treino_idx,]

dim(treino)
dim(teste)

## Regressão linear 
# y: ViolentCrimesPerPop
# x: Todos os outros

linreg <- lm(ViolentCrimesPerPop~., data=treino)
summary(linreg, digits=3)

rsquared <- Rsquared(linreg, teste, teste$ViolentCrimesPerPop)
rsquared

sprintf("R quadrado para regressão linear múltipla (crimes.csv)  é : %f", rsquared)

### 1.4c K Nearest Neighbors Regression – R( Normalized)
Ganesh, Tinniam V. Practical Machine Learning with R and Python: Machine Learning in Stereo (p. 103). Edição do Kindle. 

<<<<<<< HEAD
# Objetivo: encontrar a potência do carro (mpg) de acordo com as características ( cylinder, horsepower etc.)

=======
>>>>>>> 499b5dae9aac4d680335f9a32d5737c88cee216d
df <- read.csv("auto_mpg.csv")
head(df)
tail(df)

df1 <- as.data.frame(sapply(df, as.numeric))
head(df1)
tail(df1)

df2 <- df1 %>% select(cylinder,displacement, horsepower,weight, acceleration, year,mpg)
head(df2)
tail(df2)

df3 <- df2[complete.cases(df2),] # mantém apenas as linhas sem NA
head(df3)
<<<<<<< HEAD
tail(df3)

## Split and train
train_idx <- trainTestSplit(df3, trainPercent=75, seed=5)

train <- df3[train_idx,]
dim(train)

test <- df3[-train_idx,]
dim(test)

train.X <- train[,1:6] # feature
train.Y <- train[,7] # target

test.X <- test[,1:6] # feature
test.Y <- test[,7] # target

# Perform MinMaxScaling of feature variables (normalizar os dados para ficarem entre 0 e 1)
train.X.scaled <- MinMaxScaler(train.X)
test.X.scaled <- MinMaxScaler(test.X)

# Create a list of neighbors
rsquared <- NULL
neighbors <- c(1,2,4,6,8,10,12,15,20,25,30)

for(i in seq_along(neighbors)){
    # Fit a KNN model
    knn = knn.reg(train.X.scaled,test.X.scaled,train.Y,k=i)
    # Compute R ssquared
    rsquared[i] = knnRSquared(knn$pred,test.Y)
}

df <- data.frame(neighbors,Rsquared=rsquared)
df

# Plot the number of neighors vs the R squared
ggplot(df,aes(x=neighbors,y=Rsquared)) + 
geom_point() + 
geom_line(color="blue") + 
xlab("Number of neighbors") + 
ylab("R squared") + 
ggtitle("KNN regression - R squared vs Number of Neighors(Normalized)")

=======
tail(df3)
>>>>>>> 499b5dae9aac4d680335f9a32d5737c88cee216d

File: Chapter4-RegressionofAContinuousVariable/cap4.py
#### CAPÍTULO 4 ####
# Algoritmos de ML relacionados a regressão:
# - Regressão simples
# - Regressão múltipla
# - Regressão polinomial
# - KNN
import os
os.chdir("/Users/fernandovieira/OneDrive/1. Educacao/INFO E ESTAT/RAIZ_Livro_ML_Tinnian_Ganesh/PracticalMachineLearningWithRandPython/Chapter4-RegressionofAContinuousVariable")
os.chdir("C:/Users/70485992191/OneDrive/1. Educacao/INFO E ESTAT/RAIZ_Livro_ML_Tinnian_Ganesh/PracticalMachineLearningWithRandPython-2/Chapter4-RegressionofAContinuousVariable")
os.getcwd()
os.listdir()

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

## Ler os dados
df = pd.read_csv("Boston.csv") # Considerando o pacote "RFunctions.R", o nome do data frame sempre deve ser 'df'
df
df.columns # https://www.cs.toronto.edu/~delve/data/boston/bostonDetail.html

### REGRESSÃO SIMPLES
# Selecionar y e x
y = df["medv"]
X = df["lstat"]

## Dividir os dados entre dados de treino e de teste (75, 25)
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, random_state=0)
X_treino

# Colocando os valores em array
X_treino = X_treino.values.reshape(-1,1) 
X_teste = X_teste.values.reshape(-1,1)

X_treino.shape
X_teste.shape

## Regressão linear 
# y: Median value of owner-occupied homes in $1000's
# x: lower status of the population
linreg = LinearRegression().fit(X_treino, y_treino)
linreg.score(X_treino, y_treino)
linreg.intercept_
linreg.coef_

r2Treino = linreg.score(X_treino, y_treino)
r2Treino

r2Teste = linreg.score(X_teste, y_teste)
r2Teste

fig = plt.scatter(X_teste, y_teste) # Create a range of points. Compute yhat=coeff1*x + intercept and plot
x = np.linspace(0, 40, 20)
fig1 = plt.plot(x, linreg.coef_*x + linreg.intercept_, color="red")
fig1 = plt.title("Median value of homes ($1000) vs Lowe status (%)")
fig1 = plt.xlabel("Lower status (%)")
fig1 = plt.ylabel("Median value of owned homes ($1000)")
fig.figure.savefig('foo.png', bbox_inches='tight')
fig1.figure.savefig('foo1.png', bbox_inches='tight')
plt.show()

### REGRESSÃO MÚLTIPLA
# Ler os dados
crimesDF = pd.read_csv("crimes.csv", encoding="ISO-8859-1")
crimesDF

crimesDF.columns # https://rpubs.com/ksakwa/470184

# Remover as 7 primeiras colunas
crimesDF1 = crimesDF.iloc[:,7:crimesDF.shape[1]]

# Convert to numeric
crimesDF2 = crimesDF1.apply(pd.to_numeric, errors='coerce')

# Impute NA to 0s
crimesDF2.fillna(0, inplace=True)

# Selecionar y e x
y = crimesDF2.iloc[:,121]
X = crimesDF2.iloc[:,0:120]

## Dividir os dados entre dados de treino e de teste (75, 25)
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, random_state=0)
X_treino

## Regressão linear 
# y: Median value of owner-occupied homes in $1000's
# x: lower status of the population
linreg = LinearRegression().fit(X_treino, y_treino)
linreg.score(X_treino, y_treino)
linreg.intercept_
linreg.coef_

r2Treino = linreg.score(X_treino, y_treino)
r2Treino

r2Teste = linreg.score(X_teste, y_teste)
r2Teste

### 1.4c K Nearest Neighbors Regression – R( Normalized)
Ganesh, Tinniam V. Practical Machine Learning with R and Python: Machine Learning in Stereo (p. 105). Edição do Kindle.

# Objetivo: encontrar a potência do carro (mpg) de acordo com as características ( cylinder, horsepower etc.)

import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import MinMaxScaler

autoDF = pd.read_csv("auto_mpg.csv",encoding="ISO-8859-1")
autoDF.shape
autoDF.columns

autoDF1 = autoDF[['mpg','cylinder','displacement','horsepower','weight','acceleration','year']]

autoDF2 = autoDF1.apply(pd.to_numeric, errors='coerce')

autoDF3 = autoDF2.dropna()
autoDF3.shape

X = autoDF3[['cylinder','displacement','horsepower','weight','acceleration','year']]
y = autoDF3['mpg']

# Perform a train/ testsplit
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

# Use MinMaxScaling
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)

# Apply scaling on test set
X_test_scaled = scaler.transform(X_test)

# Create a list of neighbors
rsquared=[]
neighbors=[1,2,4,6,8,10,12,15,20,25,30]

for i in neighbors:
    # Fit a KNN model
    knnreg = KNeighborsRegressor(n_neighbors = i).fit(X_train_scaled, y_train)
    # Compute R squared
    rsquared.append(knnreg.score(X_test_scaled, y_test))
    print('R-squared test score: {:.3f}'.format(knnreg.score(X_test_scaled, y_test)))
    
# Plot the number of neighors vs the R squared
fig4=plt.plot(neighbors,rsquared)
fig4=plt.title("KNN regression - R squared vs Number of neighbors(Normalized)")
fig4=plt.xlabel("Neighbors")
fig4=plt.ylabel("R squared")
fig4.figure.savefig('foo4.png', bbox_inches='tight')
plt.show()
print("Finished plotting and saving")

## R-squared test score: 0.703
## R-squared test score: 0.810
## R-squared test score: 0.830

